### PR TITLE
Reworked identity patch for auth token

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/29096

In last upstream Google has started to block https://source.chromium.org/chromium/chromium/src/+/b4c9aeeaa5b3cfa59d6e9f1bb0d0d012f5f77ec4 getAuthToken interactive calls when idle. This breaks the old patch because
the fallback to WebAuthFlow can only be completed in interactive mode https://chromium-review.googlesource.com/c/chromium/src/+/2704571/4/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc#526.

From docs about Interactive versus silent mode (
https://developer.chrome.com/docs/apps/app_identity/#launchWebAuthFlow-interactive)

`If you invoke the API in silent mode, the API will only return a token if the provider is able to provide a token without showing any UI. This is useful in cases when an app is doing the flow at app startup, for example, or in general in cases where there is no user gesture involved.`

It means that silent mode should be supported by server side and return a token.

Samples extensions used for tests to authorize on Google servers were not authorized in silent mode and working only in interactive.

In order to fix this I force the interactive mode in fallback to WebAuthFlow each time when it is possible or explicitly requested by api, otherwise the silent mode will be used if no silent mode allowed AND it was request by api call.

It looks valid because the sense of initial chromium patch is to redirect silent modes to interactive using WebAuthFlow https://crrev.com/c/2704571

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Install https://chrome.google.com/webstore/detail/identity-api-sample/oficlfehfenioickohognhdhmmcpceil and try to login
- check token received and data can be fetched

### CODE REVIEW

File: browser/extensions/api/identity/BUILD.gn [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2023 The Brave Authors. All rights reserved.
```
Comment (bridiver) [inline browser/extensions/api/identity/BUILD.gn:1]: please move the test to api/identity/test so we don't have a directory with .h/.cc files that aren't in the BUILD.gn

File: browser/extensions/api/identity/brave_web_auth_flow.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+#include "chrome/browser/extensions/api/identity/extension_token_key.h"
+#include "chrome/browser/extensions/api/identity/identity_get_auth_token_error.h"
+#include "chrome/browser/extensions/api/identity/web_auth_flow.h"
+
+namespace extensions {
+
+class BraveWebAuthFlow : public WebAuthFlow::Delegate {
+ public:
+  using CompleteFunctionWithErrorCallback =
```
Comment (bridiver) [inline browser/extensions/api/identity/brave_web_auth_flow.h:22]: icwyu OnceCallback

File: browser/extensions/api/identity/brave_web_auth_flow.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                              const std::set<std::string>& granted_scopes)>;
+
+  BraveWebAuthFlow();
+  ~BraveWebAuthFlow() override;
+
+  static void SetTokenForTesting(const std::string& token);
+  // Used only if Google API keys aren't set up.
+  // WebAuthFlow::Delegate implementation:
+  void OnAuthFlowFailure(WebAuthFlow::Failure failure) override;
+  void OnAuthFlowURLChange(const GURL& redirect_url) override;
```
Comment (bridiver) [inline browser/extensions/api/identity/brave_web_auth_flow.h:35]: icwyu GURL

File: browser/extensions/api/identity/brave_web_auth_flow.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  // Used only if Google API keys aren't set up.
+  void StartWebAuthFlow(
+      Profile* profile,
```
Comment (bridiver) [inline browser/extensions/api/identity/brave_web_auth_flow.h:39]: icwyu Profile (forward decl)

File: browser/extensions/api/identity/brave_web_auth_flow.cc [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  token_for_testing_ = token;
+}
+
+void BraveWebAuthFlow::StartWebAuthFlow(
+    Profile* profile,
+    base::OnceClosure complete_mint_token_flow_callback,
+    CompleteFunctionWithErrorCallback complete_with_error_callback,
+    CompleteFunctionWithResultCallback complete_with_result_callback,
+    const std::string& oauth2_client_id,
+    ExtensionTokenKey token_key,
```
Comment (bridiver) [inline browser/extensions/api/identity/brave_web_auth_flow.cc:43]: icwyu ExtensionTokenKey

File: browser/extensions/api/identity/BUILD.gn (hunk 1 above)
Comment (spylogsster) [reply browser/extensions/api/identity/BUILD.gn]: done

File: browser/extensions/api/identity/brave_web_auth_flow.h (hunk 2 above)
Comment (spylogsster) [reply browser/extensions/api/identity/brave_web_auth_flow.h:26]: done

File: browser/extensions/api/identity/brave_web_auth_flow.h (hunk 3 above)
Comment (spylogsster) [reply browser/extensions/api/identity/brave_web_auth_flow.h:39]: done

File: browser/extensions/api/identity/brave_web_auth_flow.h (hunk 4 above)
Comment (spylogsster) [reply browser/extensions/api/identity/brave_web_auth_flow.h:43]: done

File: browser/extensions/api/identity/brave_web_auth_flow.cc (hunk 5 above)
Comment (spylogsster) [reply browser/extensions/api/identity/brave_web_auth_flow.cc:43]: included in .h

**Review by bridiver — APPROVED** [review]

**Pull request merged**